File: dotnet/src/utility/ValueUtility.cs
namespace BackendExample.Utility
{
    /// <summary>
    /// 值 Utility
    /// </summary>
    public class ValueUtility
    {
        /// <summary>
        /// 判断字符串是否为 null 或空字符串
        /// </summary>
        public static bool IsBlank(string? value)
        {
            return string.IsNullOrWhiteSpace(value);
        }

        //==============================================================================================================

        /// <summary>
        /// 转 bool，失败返回 null
        /// </summary>
        public static bool? ToBoolOrNull(string? value)
        {
            if (IsBlank(value))
                return null;
            if (new List<string>() { "true", "1", "t", "y", "yes", "on" }.Contains(value!.ToLower()))
                return true;
            else if (new List<string>() { "false", "0", "f", "n", "no", "off" }.Contains(value.ToLower()))
                return false;
            return null;
        }

        /// <summary>
        /// 转 bool，失败返回默认值
        /// </summary>
        public static bool ToBoolOrDefault(string? value, bool defaultValue)
        {
            bool? b = ToBoolOrNull(value);
            if (!b.HasValue)
                return defaultValue;
            return b.Value;
        }

        //==============================================================================================================

        /// <summary>
        /// 转 int，失败返回 null
        /// </summary>
        public static int? ToIntOrNull(string? value)
        {
            if (IsBlank(value))
                return null;
            if (int.TryParse(value, out int result))
                return result;
            return null;
        }

        /// <summary>
        /// 转 int，失败返回默认值
        /// </summary>
        public static int ToIntOrDefault(string? value, int defaultValue)
        {
            int? i = ToIntOrNull(value);
            if (!i.HasValue)
                return defaultValue;
            return i.Value;
        }

        //==============================================================================================================

        /// <summary>
        /// 转 long，失败返回 null
        /// </summary>
        public static long? ToLongOrNull(string? value)
        {
            if (IsBlank(value))
                return null;
            if (long.TryParse(value, out long result))
                return result;
            return null;
        }

        /// <summary>
        /// 转 long，失败返回默认值
        /// </summary>
        public static long ToLongOrDefault(string? value, long defaultValue)
        {
            long? l = ToLongOrNull(value);
            if (!l.HasValue)
                return defaultValue;
            return l.Value;
        }

        //==============================================================================================================

        /// <summary>
        /// 转 decimal，失败返回 null
        /// </summary>
        public static decimal? ToDecimalOrNull(string? value)
        {
            if (IsBlank(value))
                return null;
            if (decimal.TryParse(value, out decimal result))
                return result;
            return null;
        }

        /// <summary>
        /// 转 decimal，失败返回默认值
        /// </summary>
        public static decimal ToDecimalOrDefault(string? value, decimal defaultValue)
        {
            decimal? d = ToDecimalOrNull(value);
            if (!d.HasValue)
                return defaultValue;
            return d.Value;
        }

        //==============================================================================================================

        /// <summary>
        /// 格式化 datetime
        /// </summary>
        public static string FormatDateTime(DateTime value)
        {
            return value.ToString("yyyy-MM-dd HH:mm:ss");
        }

        /// <summary>
        /// 转 datetime，失败返回 null
        /// </summary>
        public static DateTime? ToDateTimeOrNull(string? value)
        {
            if (IsBlank(value))
                return null;
            if (DateTime.TryParse(value, out DateTime result))
                return result;
            return null;
        }

        /// <summary>
        /// 转 datetime，失败返回默认值
        /// </summary>
        public static DateTime ToDateTimeOrDefault(string? value, DateTime defaultValue)
        {
            DateTime? dt = ToDateTimeOrNull(value);
            if (!dt.HasValue)
                return defaultValue;
            return dt.Value;
        }

        //==============================================================================================================

        /// <summary>
        /// 格式化 date
        /// </summary>
        public static string FormatDate(DateOnly value)
        {
            return value.ToString("yyyy-MM-dd");
        }

        /// <summary>
        /// 转 date，失败返回 null
        /// </summary>
        public static DateOnly? ToDateOrNull(string? value)
        {
            if (IsBlank(value))
                return null;
            if (DateOnly.TryParse(value, out DateOnly result))
                return result;
            return null;
        }

        /// <summary>
        /// 转 date，失败返回默认值
        /// </summary>
        public static DateOnly ToDateOrDefault(string? value, DateOnly defaultValue)
        {
            DateOnly? d = ToDateOrNull(value);
            if (!d.HasValue)
                return defaultValue;
            return d.Value;
        }

        //==============================================================================================================

        /// <summary>
        /// 格式化 time
        /// </summary>
        public static string FormatTime(TimeOnly value)
        {
            return value.ToString("HH:mm:ss");
        }

        /// <summary>
        /// 转 time，失败返回 null
        /// </summary>
        public static TimeOnly? ToTimeOrNull(string? value)
        {
            if (IsBlank(value))
                return null;
            if (TimeOnly.TryParse(value, out TimeOnly result))
                return result;
            return null;
        }

        /// <summary>
        /// 转 time，失败返回默认值
        /// </summary>
        public static TimeOnly ToTimeOrDefault(string? value, TimeOnly defaultValue)
        {
            TimeOnly? t = ToTimeOrNull(value);
            if (!t.HasValue)
                return defaultValue;
            return t.Value;
        }
    }
}

File: python/src/utility/value_utility.py
from datetime import datetime, date, time
from decimal import Decimal


def is_blank(value) -> bool:
    """判断字符串是否为 None 或空字符串"""
    if value is None:
        return True
    s = str(value).strip()  # 去空格
    return not s


# ======================================================================================================================


# def to_string_or_empty(value) -> str:
#     """转 string，失败返回空字符串"""
#     if value is None:
#         return ""
#     return str(value).strip()  # 去空格
#
#
# def to_string_or_default(value, default: str) -> str:
#     """转 string，失败返回默认值"""
#     if value is None:
#         return default
#     return str(value).strip()  # 去空格


# ======================================================================================================================

def to_bool_or_none(value) -> bool | None:
    """转 bool，失败返回 None"""
    if value is None:
        return None
    if isinstance(value, bool):
        return value
    b = str(value).strip().lower()
    if b in ('true', '1', 't', 'y', 'yes', 'on'):
        return True
    elif b in ('false', '0', 'f', 'n', 'no', 'off'):
        return False
    return None


def to_bool_or_default(value, default: bool) -> bool:
    """转 bool，失败返回默认值"""
    b = to_bool_or_none(value)
    if b is None:
        return default
    return b


# ======================================================================================================================

def to_int_or_none(value) -> int | None:
    """转 int，失败返回 None"""
    if value is None:
        return None
    if isinstance(value, int):
        return value
    try:
        return int(value)
    except Exception:
        return None


def to_int_or_default(value, default: int) -> int:
    """转 int，失败返回默认值"""
    i = to_int_or_none(value)
    if i is None:
        return default
    return i


# ======================================================================================================================

def to_decimal_or_none(value) -> Decimal | None:
    """转 decimal，失败返回 None"""
    if value is None:
        return None
    if isinstance(value, Decimal):
        return value
    try:
        return Decimal(str(value))  # 需要 str()，否则 Decimal128 无法解析
    except Exception:
        return None


def to_decimal_or_default(value, default: Decimal) -> Decimal:
    """转 decimal，失败返回默认值"""
    d = to_decimal_or_none(value)
    if d is None:
        return default
    return d


# ======================================================================================================================

def format_datetime(value: datetime) -> str:
    """格式化 datetime"""
    return value.strftime("%Y-%m-%d %H:%M:%S")


def to_datetime_or_none(value) -> datetime | None:
    """转 datetime ，失败返回 None"""
    if value is None:
        return None
    if isinstance(value, datetime):
        return value
    try:
        # return datetime.fromisoformat(value)
        return datetime.strptime(str(value), "%Y-%m-%d %H:%M:%S")
    except Exception:
        return None


def to_datetime_or_default(value, default: datetime) -> datetime:
    """转 datetime，失败返回默认值"""
    dt = to_datetime_or_none(value)
    if dt is None:
        return default
    return dt


# ======================================================================================================================


def format_date(value: date) -> str:
    """格式化 date"""
    return value.strftime("%Y-%m-%d")


def to_date_or_none(value) -> date | None:
    """转 date ，失败返回 None"""
    if value is None:
        return None
    if isinstance(value, date):
        return value
    try:
        return datetime.strptime(str(value), "%Y-%m-%d").date()
    except Exception:
        return None


def to_date_or_default(value, default: date) -> date:
    """转 date，失败返回默认值"""
    d = to_date_or_none(value)
    if d is None:
        return default
    return d


# ======================================================================================================================


def format_time(value: time) -> str:
    """格式化 time"""
    return value.strftime("%H:%M:%S")


def to_time_or_none(value) -> time | None:
    """转 time ，失败返回 None"""
    if value is None:
        return None
    if isinstance(value, time):
        return value
    try:
        return datetime.strptime(str(value), "%H:%M:%S").time()
    except Exception:
        return None


def to_time_or_default(value, default: time) -> time:
    """转 time，失败返回默认值"""
    t = to_time_or_none(value)
    if t is None:
        return default
    return t
